src/vision/status_quo/barbara_tries_unix_socket.md

```diff
@@ -0,0 +1,314 @@
+    }
+}
+                                 
+impl<S: AsyncRead + AsyncWrite + Unpin> AsyncWrite for TracingStream<S> {
+    fn poll_flush(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Result<(), std::io::Error>> {
+        Pin::new(&mut self.inner).poll_flush(cx)
+    } 
+    
+    fn poll_shutdown(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Result<(), std::io::Error>> {
+        Pin::new(&mut self.inner).poll_shutdown(cx)
+    }
+ 
+    fn poll_write(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &[u8],
+    ) -> Poll<Result<usize, std::io::Error>> {
+        let poll_result = Pin::new(&mut self.inner).poll_write(cx, buf);
+        for line in String::from_utf8_lossy(buf).into_owned().lines() {
+            println!("< {}", line);
+        }
+        poll_result
+    }
+}
+```
+
+Barbara happily copy pasted the code and wrap the `stream` within `TracingStream`. Running it with logs gives (same thing, in some cases it works, in some cases it did not work):
+```
+ TRACE mio::poll > registering event source with poller: token=Token(0), interests=READABLE | WRITABLE
+connected
+< GET / HTTP/1.1
+< 
+> HTTP/1.1 200 OK
+> content-type: text/plain; charset=utf-8
+> content-length: 12
+> date: Tue, 04 May 2021 17:02:49 GMT
+> 
+> Hello World!
+Some(Ok(b"Hello World!"))
+sending 2
+ TRACE want      > signal: Want
+ TRACE want      > signal: Want
+ TRACE mio::poll > deregistering event source from poller
+ TRACE want      > signal: Closed
+closed
+Error: hyper::Error(Canceled, "connection was not ready")
+```
+
+Barbara thought this probably only affects a unix socket but nope, even swapping it back with TCP socket does not work either. Now, not just Barbara was confused, even the other developers who offered help was confused now.
+
+## The single magical line
+
+After some time, a developer found a solution, just a single line. Barbara added the line and it works like a charm but it still feels like magic.
+
+```rust
+use futures::future;
+
+    // this new line below was added for second request
+    future::poll_fn(|cx| request_sender.poll_ready(cx)).await?;
+    let request = Request::get("/").body(Body::empty())?;
+    println!("sending 2");
+    let mut response = request_sender.send_request(request).await?;
+    println!("sent 2");
+    println!("{:?}", response.body_mut().data().await);
+```
+
+Barbara still have no idea why it needs to be done this way. But at least it works now.
+
+## 🤔 Frequently Asked Questions
+
+*Here are some standard FAQ to get you started. Feel free to add more!*
+
+### **What are the morals of the story?**
+
+Barbara is not able to see the problem right away. Usually missing an `await` is an issue but in this case, not awaiting on another variable or not polling for ready when using a low-level library may the program incorrect, it is also hard to debug and figure out what is the correct solution.
```

Comment:
For me personally, this "solution" (adding the call to `future::poll_fn`) feels very unsatisfying!  Do you think Barbara felt the same way?  Did she feel like this was really the solution to her problem?  Or did it feel more like some workaround/bandaid?

Reply:
Yeah, feels like a magic to me, I wonder why it wasn't mentioned or like this is so implicit. It makes doing `async` no difference than `unsafe`, where I had to manually audit which block if they have completed. Barbara would probably feel the same way if she had no idea it is required in the first place. At least this part I think could be even marked `unsafe` since it is sort of "undefined behavior" in async world (not what the original spec says). At least to me I think this is understandable for a low-level library but it would be better if it is documented or marked as unsafe, at least users won't be confused, not sure if this can be embedded in the type system so it can be safe.

Reply:
I would like to see some of *this* in the morals! "Feels like magic" and having to manually audit. That exact analogy to unsafe has occurred to me as well, reading some of the status quo stories.

Reply:
```suggestion
Barbara is not able to see the problem right away. Usually missing an `await` is an issue but in this case, not awaiting on another variable or not polling for ready when using a low-level library may the program incorrect, it is also hard to debug and figure out what is the correct solution.

In a way, some of the fixes "feels like magic". Sometimes polling is required to be done but where? It may make people afraid of using `async/.await` and end up writing safety net code (for example, writing code to do type checking in weakly typed languages in every lines of code to be safe).

Having these pitfalls in mind, one can easily relate it back to unsafe. If there are unsafe blocks, the user needs to manually audit every specific code block for undefined behaviors. But in the case of async, the situation is someone similar such that the user need to audit the whole async code blocks (which is a lot compared to unsafe) for "undefined behaviors", rather than having when it compiles it works sort of behavior.
```